### ISSUE TITLE
Crash after hitting AI HQ

### ISSUE THREAD

**Issue by alasiax** [widelands/widelands#5050]
**Describe the bug**
The game consistently crashes when I hit the grey AI HQ in Astoria R2 map. I was having this problem in git version 25438 which does not have a debug version, so I downloaded the Version 1.1~git25441[d355de6@HEAD] (Debug) and started it from the command line. 
I've been able to reliably reproduce this from several saves made at different points (ie I went back and loaded an earlier autosave, to see if it happened again if I did things differently) - it doesn't matter whether I go off and do other things first, when I hit the Grey/Player 5 HQ, it crashes. The game was started in 25438, which was where I noticed it. I've attached two sets of files, but don't know if the crashes are related because with the debug version I got two different results. Hopefully I've grabbed the right files. 

_In game 1:_
Crash shows Microsoft Visual C++ runtime error, Assertion Failed

line 1014 in ../src/ui_basic/panel.cc 
`
!mousegrab_ || mousegrab_ == this`

(error screencap attached.)
For game 1 the stderr.txt file is empty. 

I have seen this error before, in multiplayer I think on the same map. The error is definitely the same as I took a screenshot of it. I did not properly note the WL version number unfortunately, but I think it was git 25417 as it was a debug version, and I think that's the only one I've installed prior to the current one.

_In game 2_
I tried loading from an earlier save point.
After some long time playing (and avoiding the grey HQ), it still crashed when I hit it. This time there was a stderr.txt output, but the game simply stopped. Actually, I thought that I had avoided the crash as I saw the HQ blow up, but it crashed very shortly after.
stderr.txt:
```
terminate called after throwing an instance of 'WException'
  what():  [../src/logic/playersmanager.cc:93] Player End status for player 5 already reported
```

**To Reproduce**
1. Attack grey player HQ

**Expected behavior**
I expect chaos and mayhem as the Amazons achieve victory over the Barbarian horde. This happened in game 2, but then the game crashes shortly after.

**Version:**
Widelands Version 1.1~git25441[d355de6@HEAD] (Debug), also occurs on 25438 (non-debug).

**Additional context**
I'm using Windows 10 Laptop, with a generic USB optical mouse.
Intel Core i5 with 16GB RAM.

Edit: I made sure I had the matching files. 
[game 2.zip](https://github.com/widelands/widelands/files/7072779/game.2.zip)
[game 1.zip](https://github.com/widelands/widelands/files/7072781/game.1.zip)

**Comment by Noordfrees** [comment]
In game1 I also get the playersmanager assertion. gdb shows that the result was indeed added long ago:
```
(gdb) p players_end_status_
$1 = std::vector of length 2, capacity 2 = {{player = 5 '\005', result = Widelands::PlayerEndResult::kLost, time = {value_ = 26438300}, 
    info = "win_condition=HQ Hunter;win_condition_version=2;tribe=barbarians"}, {player = 2 '\002', result = Widelands::PlayerEndResult::kLost, time = {value_ = 39904700}, 
    info = "win_condition=HQ Hunter;win_condition_version=2;tribe=empire"}}

(gdb) p status
$3 = (const Widelands::PlayerEndStatus &) @0x7fffbdbedc70: {player = 5 '\005', result = Widelands::PlayerEndResult::kLost, time = {value_ = 45091300}, 
  info = "win_condition=HQ Hunter;win_condition_version=2;tribe=barbarians"}
```

So player 5 lost at time 26438300 (7:20:38) and now again at 45091300 (12:31:31). Perhaps a bug in the win condition scripting? 

In the replay this end status is already added, do you have an earlier one?

---

The mousegrab_ assert looks like a separate bug in UI code, steps to reproduce would be good

**Comment by alasiax** [comment]
These were created with 25438. Replay starts at 7.07. I have others if you need them.
[game 7.07.zip](https://github.com/widelands/widelands/files/7072904/game.7.07.zip)

Re: the mouse issue, it literally happened the same as the HQ hit today, which is why I wasn't sure if there's a connection. I will look for the multiplayer game files in the morning, but I think I was attacking when that happened too, but I don't recall if I was hitting a HQ. That was a couple of weeks ago. Both times the error has flashed up I was using the debug version. 

Edit: ie, the game crashed during my hit on the grey HQ, and as it was the first one I did with the debug version, I didn't know if it was now producing an actual reportable error rather than simply crashing the without warning.

**Comment by alasiax** [comment]
Excellent news! I've been able to make the error happen on a different map. Hooray for reproducible crashes? 

Edit: The win condition in this one is Autocrat. This map isn't super conducive to a quick restart, so I don't know if the error would be reproducible under the HQ Hunter condition. 

I don't remember if the earlier game1/2 crashed when I hit the red player HQ in the previous games, but I don't think so. I'll try going after another player in the current game (River Explorers) and see if it happens elsewhere.

Files attached.
[game3-riverexplorers.zip](https://github.com/widelands/widelands/files/7076450/game3-riverexplorers.zip)

**Comment by alasiax** [comment]
Ok, the game doesn't crash when I take out HQ and the last warehouse of the green player in the river explorers game.  I avoided going too close to the yellow HQ this time, in case that was the problem. It also crashed regardless of whether I conquered military buildings within the HQ sphere of control before attacking the HQ itself.

[game3-riverexplorers.v.1.zip](https://github.com/widelands/widelands/files/7076845/game3-riverexplorers.v.1.zip)

**Comment by alasiax** [comment]
> The mousegrab_ assert looks like a separate bug in UI code, steps to reproduce would be good

As this happened again today with this error, I'll start a new bug report.

**Comment by klaus-halfmann** [comment]
mousegrab_ assert happend for me, too. Try to click often between widelands and some other windows.
With and without Fullscreen mode (e.g. Switch vial ALT-TAB). I would assuem with the new Multithreading
this may lead to events where the Window does not have the correct mouse/window status, yet.

Should not be related (directly) to actually building something or other other in game actions.
Mabe when (after starting to build something) you switch to some other window?

**Comment by alasiax** [comment]
Thanks for raising that it might be able to be triggered by window switching @klaus-halfmann. I will look out for that in future. I'm linking further notes on this to #5056 in case it helps there.

**Comment by alasiax** [comment]
Two more sets of files (on the twinkling waves map)
Widelands Version 1.1~git25441[d355de6@HEAD] (Debug)

The game4 - desync is a replay of a game which did not immediately crash on HQ takeover. Interestingly, I closed the game naturally, but the replay is throwing a desync error, although it was definitely started in the same version. There were also warnings in the replay stdout.txt 
As I exited the game withou issue, unfortunately I did not think to check the stdout.txt when I closed the game.

game4 was loaded from an autosave of the above game prior to conquering the HQ. In this case the error about the Player End reported is produced when the game crashes.

```
terminate called after throwing an instance of 'WException'
  what():  [../src/logic/playersmanager.cc:93] Player End status for player 3 already reported
```

[game4 - twinkling waves desync.zip](https://github.com/widelands/widelands/files/7097586/game4.-.twinkling.waves.desync.zip)
[game4 - twinkling waves.zip](https://github.com/widelands/widelands/files/7097588/game4.-.twinkling.waves.zip)
